### PR TITLE
Renaming Biosemicaps to 10-10 EEG system 

### PR DESCRIPTION
**Pull request by NemaVatsala**
## Description
Process and context menu to rename EEG channels from the BioSemi naming system to the standard 10-10 system. Renaming only happens for 16-, 32- and 64-electrode caps as these have a one to one correspondence. See [BioSemi caps documentation](https://www.biosemi.com/headcap.htm).

Renaming happens if:
 - There are EEG channels, AND
 - Current channels names correspond to BioSemi style: A1...A32, B1...B32 and so on
 - All EEG channels make up a Biosemi cap. No missing or extra EEG channels

Context menu, only appears for Channel file with `BDF` in their names.

## Result:

| Before  | After |
|--------|-------|
| image | image |

Edited by @rcassani

### CODE REVIEW

**Comment by rcassani** [comment]
@NemaVatsala, I have commented out the renaming code in `tree_callbacks` and prepare the template for perform the renaming of channels in a process (`process_rename_biosemi`). 

The code has various `TODO` parts that need to be completed with commented code from `tree_callbacks`

At this moment the process is functional and it will only add an entry in the `History` field of the channel file. It can be found in `Run > Import > Channel file > Rename EEG channels Biosemi caps`. 
And it can be also run as:

    process_channel_biosemi('Compute', channelRelativeFileName), 

which will be they way it is executed from `tree_callbacks`


image

File: toolbox/process/functions/process_channel_biosemi.m [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+            
+            % Loop through each channel name
+            for i = 1:length(ChannelMat.Channel)
+                originalName = channelData{i};
+                
+                % Check if the channel has a mapping in the dictionary
+                if isKey(channelMap, originalName)
+                    newName = channelMap(originalName);
+                    renamedChannels{i} = newName;  % Rename the channel
+                end
```
Comment (rcassani) [inline toolbox/process/functions/process_channel_biosemi.m:157]: This check should be done before calling `Compute()`

File: toolbox/process/functions/process_channel_biosemi.m [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+        
+        
+   
+            
+    
+
+    %       - All EEG channels make up a Biosemi cap, no 1 channel more, no 1 channel less.
+
+        missingElectrodes = setdiff(keys(Tentensystem), channels);
+        extraElectrodes = setdiff(channels, keys(Tentensystem));
```
Comment (rcassani) [inline toolbox/process/functions/process_channel_biosemi.m:91]: `extraChannels` is ok as Biosemi systems can record other signals as ECG, respiration, etc. We just need to check that all the channels in the cap are in the recording. What it is not ok is extra EEG channels

File: toolbox/process/functions/process_channel_biosemi.m [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+                % Check if the channel has a mapping in the dictionary
+                if isKey(channelMap, originalName)
+                    newName = channelMap(originalName);
+                    renamedChannels{i} = newName;  % Rename the channel
+                end
+            end
+        end
+
+
+    ChannelMatNew.Channel = ChannelMatOld.Channel;
```
Comment (rcassani) [inline toolbox/process/functions/process_channel_biosemi.m:162]: `ChannelMatNew` should be an structure with the `Channel` and `History` fields. However at this point `ChannelMatNew` is a cell array with the renamed channels

File: toolbox/process/functions/process_channel_biosemi.m [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+
+        mapping32 = containers.Map({'A1', 'A30','A2','A29', 'A3','A4' 'A31', 'A27', 'A28', 'A6','A5','A26', 'A25','A7','A8','A32', 'A23', 'A24', 'A10', 'A9','A22', 'A21','A11','A12', 'A13','A19', 'A20', 'A14','A18','A15', 'A16','A17'}, {'Fp1','Fp2', 'AF3','AF4', 'F7','F3', 'Fz','F4', 'F8', 'FC5','FC1','FC2','FC6', 'C3', 'Cz', 'C4', 'T8', 'CP5', 'CP1','CP2','CP6','P7','P3','Pz','P4' 'P8', 'PO3','PO4','O1', 'Oz','O2'});
+      
+    % TODO: Generate one-to-one channel name maps for Biosemi caps
+    %       Maps can be as simple as cell arrays size (nChannels, 2), one column Biosemi, one column 10-10 system
+    %       Then comparisons are simpler
+    %       Or they can be containers.Map objects- more flexible
+
+    % Create a dictionary-like structure for mapping
+       if length(ChannelMat.Channel) == 32
```
Comment (rcassani) [inline toolbox/process/functions/process_channel_biosemi.m:185]: ChannelMat is not defined in this function.
The check for the mapping should depend only on EEG-Type channels, not the total number

File: toolbox/process/functions/process_channel_biosemi.m [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+    % Update file
+    bst_save(ChannelFileFull, ChannelMatNew, 'v7', 1);
+    ChannelFileNew = file_short(ChannelFileFull);
+end
+
+%% ===== NAME MAPS =====
+function ChannelMap = GetBiosemiMaps()
+%Biosemi structure can be found at: https://www.biosemi.com/headcap.htm
+    %ChannelNameMaps = [];
+        mapping64 = containers.Map({'A1', 'B1', 'B2', 'A2', 'A3', 'B5', 'B4', 'B3', 'A7', 'A6', 'A5', 'A4', 'B6', 'B7', 'B8', 'B9', 'B10', 'A8', 'A9', 'A10', 'A11', 'B15', 'B14', 'B13', 'B12', 'B11', 'A15', 'A14', 'A13', 'A12', 'B16', 'B17', 'B18', 'B19', 'B20', 'A16', 'A17', 'A18', 'A19', 'A32', 'B24', 'B23', 'B22', 'B21', 'A24', 'A23', 'A22', 'A21', 'A20', 'A31', 'B25', 'B26', 'B27', 'B28', 'B29', 'A25', 'A26', 'A30', 'B31', 'PO8', 'A27', 'A29', 'B32', 'A28'}, {'Fp1', 'Fpz', 'Fp2', 'AF7', 'AF3', 'AFz', 'AF4', 'AF8', 'F7', 'F5', 'F3', 'F1', 'Fz', 'F2', 'F4', 'F6', 'F8', 'FT7', 'FC5', 'FC3', 'FC1', 'FCz', 'FC2', 'FC4', 'FC6', 'FT8', 'T7', 'C5', 'C3', 'C1', 'Cz', 'C2', 'C4', 'C6', 'T8', 'TP7', 'CP5', 'CP3', 'CP1', 'CPz', 'CP2', 'CP4', 'CP6', 'TP8', 'P9', 'P7', 'P5', 'P3', 'P1', 'Pz', 'P2', 'P4', 'P6', 'P8', 'P10', 'PO7', 'PO3', 'POZ', 'PO4', 'O1', 'OZ', 'O2', 'IZ'});
```
Comment (rcassani) [inline toolbox/process/functions/process_channel_biosemi.m:175]: For sake or readability divided this line in more

File: toolbox/process/functions/process_channel_biosemi.m [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+
+        mapping32 = containers.Map({'A1', 'A30','A2','A29', 'A3','A4' 'A31', 'A27', 'A28', 'A6','A5','A26', 'A25','A7','A8','A32', 'A23', 'A24', 'A10', 'A9','A22', 'A21','A11','A12', 'A13','A19', 'A20', 'A14','A18','A15', 'A16','A17'}, {'Fp1','Fp2', 'AF3','AF4', 'F7','F3', 'Fz','F4', 'F8', 'FC5','FC1','FC2','FC6', 'C3', 'Cz', 'C4', 'T8', 'CP5', 'CP1','CP2','CP6','P7','P3','Pz','P4' 'P8', 'PO3','PO4','O1', 'Oz','O2'});
```
Comment (rcassani) [inline toolbox/process/functions/process_channel_biosemi.m:177]: Same comment, long line

File: toolbox/process/functions/process_channel_biosemi.m
Comment (rcassani) [inline toolbox/process/functions/process_channel_biosemi.m:1]: Better indentation and removing empty lines would improve readability

**Review by rcassani — COMMENTED** [review]
The development of the process is moving forward.
There are comments that need to be solved with changes.

File: toolbox/process/functions/process_channel_biosemi.m (hunk 2 above)
Comment (NemaVatsala) [reply toolbox/process/functions/process_channel_biosemi.m]: I have taken both the case of missing and Extra eeg channels in this iteration of the commit

File: toolbox/process/functions/process_channel_biosemi.m [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+        %       - If all conditions are met, then rename channels to 10-10 system according Biosemi caps.
+                    %if not, return the missing channels                   
+                  
+        % Display the results
+
+        if isempty(missingElectrodes) && isempty(extraElectrodes)
+                            % Channel data and channel map
+                                                       
+                            % Call the renameChannels function
+                            newChannels = renameChannels(ChannelMat, channelMap);
```
[the remark this answers is not in the document]
Comment (NemaVatsala) [reply toolbox/process/functions/process_channel_biosemi.m]: Addressed in this iteration

File: toolbox/process/functions/process_channel_biosemi.m [hunk 8]
```diff
@@ -0,0 +1,194 @@
...
+                        else
+                            if ~isempty(missingElectrodes)
+                                
+                                disp(['Missing electrode numbers:' num2str(missingElectrodes)]);
+                            end
+                            if ~isempty(extraElectrodes)
+                               
+                                disp(['Extra electrode numbers:' num2str(extraElectrodes)]);
+                            end
```
[the remark this answers is not in the document]
Comment (NemaVatsala) [reply toolbox/process/functions/process_channel_biosemi.m]: Tried to Address in this iteration

File: toolbox/process/functions/process_channel_biosemi.m [hunk 9]
```diff
@@ -0,0 +1,194 @@
...
+            for i = 1:length(ChannelMat.Channel)
+                originalName = channelData{i};
+                
+                % Check if the channel has a mapping in the dictionary
+                if isKey(channelMap, originalName)
+                    newName = channelMap(originalName);
+                    renamedChannels{i} = newName;  % Rename the channel
+                end
+            end
+        end
```
[the remark this answers is not in the document]
Comment (NemaVatsala) [reply toolbox/process/functions/process_channel_biosemi.m]: I have changed it now

File: toolbox/process/functions/process_channel_biosemi.m (hunk 3 above)
Comment (rcassani) [reply toolbox/process/functions/process_channel_biosemi.m:150]: This is the structure for `ChannelMat`
https://neuroimage.usc.edu/brainstorm/Tutorials/ChannelFile#On_the_hard_drive

File: toolbox/process/functions/process_channel_biosemi.m
Comment (NemaVatsala) [reply toolbox/process/functions/process_channel_biosemi.m:1]: Addressed in the [latest commit](https://github.com/brainstorm-tools/brainstorm3/blob/512c81ffd91d545b271902b75188c4aed978bdd4/toolbox/process/functions/process_channel_biosemi.m)

File: toolbox/process/functions/process_channel_biosemi.m [hunk 10]
```diff
@@ -0,0 +1,194 @@
...
+    % ===== GET CHANNEL FILE  =====
+    sChannel = bst_get('ChannelForStudy', [sInputs.iStudy]);
+
+    % ===== VERIFY DEVICE AND BIOSEMI NAMING SYSTEM =====
+    % Load channel file
+    ChannelMat = in_bst_channel(sChannel.FileName, 'Channel');
+    ChannelMat.Channel;
+    % TODO: Check:
+
+    %       - There are EEG channels, AND
```
[the remark this answers is not in the document]
Comment (NemaVatsala) [reply toolbox/process/functions/process_channel_biosemi.m:68]: This has been addressed in the [latest commit](https://github.com/brainstorm-tools/brainstorm3/blob/512c81ffd91d545b271902b75188c4aed978bdd4/toolbox/process/functions/process_channel_biosemi.m)

File: toolbox/tree/tree_callbacks.m [hunk 11]
```diff
@@ -895,7 +895,92 @@
...
+%                             disp(result)
+%                         end
+%                         if ~isempty(extraElectrodes)
+%                             result=extraElectrodes;
+%                             disp(result)
+%                         end
+%                     end
+%                 end
+% 
+%
```
[the remark this answers is not in the document]
Comment (NemaVatsala) [reply toolbox/tree/tree_callbacks.m]: Done.

**Comment by rcassani** [comment]
@NemaVatsala FYI

**Pull request merged**